### PR TITLE
Adding new csvToJSONString scalar function

### PR DESCRIPTION
**Pull request by ollidraese**
New scalar function `csvToJSONString` to convert CSV strings into JSON strings.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New `csvToJSONString` scalar function.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by ollidraese** [comment]
This function should also be cherry-picked into upstream!

File: tests/queries/0_stateless/03442_csvToJSONString.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
+DROP TABLE IF EXISTS csv_to_json_test;
+
+CREATE TABLE
+    csv_to_json_test (entry Int32, csv String) ENGINE = Memory;
```
Comment (dcheckoway) [inline tests/queries/0_stateless/03442_csvToJSONString.sql:4]: Can you make this `Nullable(String)` and insert a `NULL` in there for fun?  Wanna make sure null in, null out, with no :boom:

File: docs/en/sql-reference/functions/json-functions.md [hunk 2]
```diff
@@ -1109,6 +1109,64 @@ Result:
...
+
+***Supported options for third parameter***
+- `delimiter` - The delimiter character (default is `,`)
+- `quote` - Limit the quote character to either `"` or `'` (default is allowing both `"` and `'`)
+- `null` - The null representation (default is `\N`)
+- `autodetect_types` - Decides if booleans and numerics are automatically detected and stored as native type in JSON (default is `true`)
+
+**Returned value**
+
+- JSON representation of the CSV data. [String](../data-types/string.md).
```
Comment (dcheckoway) [inline docs/en/sql-reference/functions/json-functions.md:1150]: And if it supports null-in-null-out, it's worth noting that here.

File: tests/queries/0_stateless/03442_csvToJSONString.sql [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    (5, 'Key5|17|Rudolf'),
+    (6, '"The,key,with,quotes","The value with,quotes",",;''"');
+
+SELECT csvToJSONString ('id,property,name', csv) AS json FROM csv_to_json_test WHERE entry < 5 ORDER BY entry;
+SELECT csvToJSONString ('id|property|name', csv, 'delimiter="|"') AS json FROM csv_to_json_test WHERE entry == 5;
+SELECT csvToJSONString ('id|property|name', csv, 'delimiter="|",autodetect_types="false"') AS json FROM csv_to_json_test WHERE entry == 5;
+SELECT csvToJSONString (',property,name', csv) AS json FROM csv_to_json_test WHERE entry < 3 ORDER BY entry;
+SELECT csvToJSONString ('id,,name', csv) AS json FROM csv_to_json_test WHERE entry < 3 ORDER BY entry;
+SELECT csvToJSONString ('id,property', csv) AS json FROM csv_to_json_test WHERE entry < 3 ORDER BY entry;
+SELECT csvToJSONString ('id,property,name', csv) AS json FROM csv_to_json_test WHERE entry == 6;
```
Comment (dcheckoway) [inline tests/queries/0_stateless/03442_csvToJSONString.sql:22]: What happens if you pass in 4 field names but there are only 3 in the string?

What happens if you pass empty string `''` for `csvString`?

Comment (dcheckoway) [reply tests/queries/0_stateless/03442_csvToJSONString.sql:22]: What happens if you pass invalid CSV, i.e. unterminated quoted value like
```
Key7,42,"this string is not terminated
```

Comment (dcheckoway) [reply tests/queries/0_stateless/03442_csvToJSONString.sql:22]: One more thing to test...escaped quotes, i.e.
```
Key8,42,"this has ""escaped"" quotes"
```

File: docs/en/sql-reference/functions/json-functions.md [hunk 4]
```diff
@@ -1109,6 +1109,62 @@ Result:
...
+enabled, a string "42" would be stored as numeric in the JSON output.
+
+The standard [CSV format settings](../../interfaces/formats/CSV/CSV.md) apply to the CSV parsing 
+of this function.
+
+**Syntax**
+
+```sql
+csvToJSONString(fieldNames, csvString)
+csvToJSONString(fieldNames, csvString, autodetect_types)
```
Comment (dcheckoway) [inline docs/en/sql-reference/functions/json-functions.md:1137]: Nit: camel instead of snake case on the 3rd arg

File: src/Functions/csvToJsonString.cpp [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+                    }
+                }
+                else
+                {
+                    json.set(key, value);
+                }
+            }
+        }
+
+        dest.str("");
```
Comment (dcheckoway) [inline src/Functions/csvToJsonString.cpp:133]: Curious why this was necessary?

File: tests/queries/0_stateless/03442_csvToJSONString.sql [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+----------------------------------------------------------------------------------------------
+-- second, try with nullable string input
+----------------------------------------------------------------------------------------------
+CREATE TABLE csv_to_json_test_nullable (entry Int32, csv Nullable(String)) ENGINE = Memory;
+
+INSERT INTO csv_to_json_test_nullable
+VALUES
+    (1, 'Field1.0,Field1.1,\\N'),
+    (2, 'Field2.0,\\N,Field2.2'),
+    (3, '\\N,Field3.1,Field3.2'),
```
Comment (dcheckoway) [inline tests/queries/0_stateless/03442_csvToJSONString.sql:48]: Can you add coverage of empty string?  i.e. `,\N,` which I think yields just `{"col2":null}` (and the other columns aren't present).

File: src/Functions/csvToJsonString.cpp [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        std::stringstream & dest,
+        bool detectTypes)
+    {
+        const auto n = std::min(fieldNames.size(), fieldValues.size());
+        Poco::JSON::Object json;
+
+        for (size_t i = 0; i < n; ++i)
+        {
+            if (fieldNames.at(i).empty())
+                continue;
```
Comment (dcheckoway) [inline src/Functions/csvToJsonString.cpp:95]: TBH this feels like it deserves a CSV format setting (which doesn't exist, I don't think), i.e. "treat empty as null."  So you could decide if you want a literal `""` empty string or null.

File: src/Functions/csvToJsonString.cpp (hunk 5 above)
Comment (dcheckoway) [reply src/Functions/csvToJsonString.cpp:133]: Oh, it's because it's being reused, and this is just "clear" right?

File: src/Functions/csvToJsonString.cpp [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "First argument (field names) of function {} must be String constant, got {}",
+                getName(),
+                arguments[0]->getName());
+        }
+
+        // Second argument must be (nullable) string (CSV values)
+        which = WhichDataType(arguments[1]);
+        bool isCSVNullable = which.isNullable();
+        if (!which.isString() && !isCSVNullable)
```
Comment (dcheckoway) [inline src/Functions/csvToJsonString.cpp:181]: I'm still getting to know the type system, but if we see it's Nullable, do we need to check the inner/nested type?  i.e. could it be `Nullable(UInt32)`?

File: src/Functions/csvToJsonString.cpp [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+                detectTypes = constOptions->getValue<UInt8>() != 0;
+            }
+        }
+
+        // we expect the field names to be a string constant
+        if (const auto * constFieldNames = checkAndGetColumnConst<ColumnString>(arguments[0].column.get()); constFieldNames)
+        {
+            const auto fieldNamesStr = constFieldNames->getDataAt(0);
+            std::vector<std::string> fieldNames;
+            fieldNames.reserve(128);
```
Comment (dcheckoway) [inline src/Functions/csvToJsonString.cpp:235]: Can you shed light on why 128 (vs. 42 or anything else)?  May be worth a preprocessor constant since we have a few places where 128 is used (assuming they're intended to be aligned).

File: src/Functions/csvToJsonString.cpp (hunk 7 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:95]: This is doing something different (as described in the doc). It allows you to omit known fields in the output:
If fieldNames is `field1,,field3`, then field2 is not generated in the output. This is field names, not data

File: src/Functions/csvToJsonString.cpp (hunk 5 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:133]: yes, The destination stringstream is not recreated between single rows and therefore needs to be reset to empty string.

File: src/Functions/csvToJsonString.cpp (hunk 8 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:181]: Good find. Fixes with next commit.

File: tests/queries/0_stateless/03442_csvToJSONString.sql (hunk 6 above)
Comment (ollidraese) [reply tests/queries/0_stateless/03442_csvToJSONString.sql:48]: done, next commit

File: src/Functions/csvToJsonString.cpp [hunk 10]
```diff
@@ -42,8 +38,8 @@ namespace
   * - All other values are treated as strings
   * 
   * Example:
-  * csvToJSONString('name,age', 'John,42') = '{"name":"John","age":42}'
-  * csvToJSONString('name|age', 'John|42', ) SETTINGS format_csv_delimiter = '|' = '{"name":"John","age":42}'
+  * csvToJSONString('name,age', 'John,42') => '{"name":"John","age":42}'
```
Comment (dcheckoway) [inline src/Functions/csvToJsonString.cpp:41]: Nit ⛏️  did you mean for the `>` to sneak in here?  If so, maybe use it on the next example as well for consistency.

File: tests/queries/0_stateless/03442_csvToJSONString.sql [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    (4, NULL);
+
+-- ensure that null-values in CSV (\\N) are detected as well as full CSV can be NULL
```
Comment (itsalvinho) [inline tests/queries/0_stateless/03442_csvToJSONString.sql:54]: Maybe worth adding a test for a custom null representation as well? eg. `format_csv_null_representation='null'`

File: src/Functions/csvToJsonString.cpp [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    explicit FunctionCsvToJsonString(ContextPtr context_)
+        : context(std::move(context_))
+    {
+    }
+    static FunctionPtr create(ContextPtr context) { return std::make_shared<FunctionCsvToJsonString>(context); }
+
+    String getName() const override { return name; }
+
+    // this function supports two or three arguments
+    size_t getNumberOfArguments() const override { return 0; }
```
Comment (itsalvinho) [inline src/Functions/csvToJsonString.cpp:147]: Is it alright to return 0 here if we support 2-3 args?

File: src/Functions/csvToJsonString.cpp (hunk 10 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:41]: Yes, will add the '>'. I found the two equals (due to settings) otherwise confusing.

File: src/Functions/csvToJsonString.cpp [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+            }
+
+            // col:1 (csv input) is neither String nor ConstString: must be NULL Constant -> return NULL column
+            auto null_col = ColumnNullable::create(ColumnString::create(), ColumnUInt8::create());
+            null_col->insertDefault();
+            return ColumnConst::create(std::move(null_col), input_rows_count);
+        }
+
+        // we should never get here as the framework ensures that col:0 and col:2 are constant
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Function {} failed to validate arguments", getName());
```
[the remark this answers is not in the document]
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:287]: No, this one is already known through include (from src/Common/typeid_cast.h)

File: src/Functions/csvToJsonString.cpp (hunk 12 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:147]: From IFunction:
`/// For non-variadic functions, return number of arguments; otherwise return zero (that should be ignored).`
This is a variadic function.

**Comment by itsalvinho** [comment]
As mentioned in the walkthrough call, to be able to support the pushdown for `extract type=delim` and its options `escapechar`, `delimChar`, `quoteChar`, and `nullValue`, I have two concerns:

1. Users can have many extracts with different options specified each time in one query. Using `SETTINGS` may be an issue if it applies to the entire query. It'd okay if it can be applied on the sub-query level we may have to consider adding options as arguments to the function itself instead.

2. It doesn't seem like `escapeChar` and `quoteChar` are configurable `CSV` format `SETTINGS`. I did find `format_custom_escaping_rule` ([here](https://clickhouse.com/docs/operations/settings/formats#format_custom_escaping_rule)) but that's not `CSV` specific, looks like it's a `CustomSeparated` format thing. Feels weird to mix these settings if we choose to do this. And you mentioned the built-in splitCSV function might not even support those functions, which sucks.

Feels like we'd may need a separate implementation for parsing "custom separated" strings, let me know what you think.

File: tests/queries/0_stateless/03442_csvToJSONString.sql (hunk 11 above)
Comment (ollidraese) [reply tests/queries/0_stateless/03442_csvToJSONString.sql:54]: done

**Comment by ollidraese** [comment]
@itsalvinho I added back the function-level options (the query context is still providing the defaults). Also added support for the quote character.

File: docs/en/sql-reference/functions/json-functions.md [hunk 14]
```diff
@@ -1109,6 +1109,62 @@ Result:
 - [output_format_json_quote_denormals](/operations/settings/formats#output_format_json_quote_denormals)
 
 
+### csvToJSONString {#csvtojsonstring}
```
Comment (cdahl-dev) [inline docs/en/sql-reference/functions/json-functions.md:1112]: What do we think about also having a function that converts csv to a Map? We may see better performance than building a JSON string.

File: docs/en/sql-reference/functions/json-functions.md [hunk 15]
```diff
@@ -1109,6 +1109,62 @@ Result:
...
+
+```sql
+SELECT csvToJSONString('name,age', 'John,42');
+SELECT csvToJSONString('name|age', 'Clark|15') SETTINGS format_csv_delimiter='|';
+SELECT csvToJSONString('name$age', 'Clark$15', 'format_csv_delimiter="$",detect_types=false');
+```
+
+Result:
+
+```text
```
Comment (cdahl-dev) [inline docs/en/sql-reference/functions/json-functions.md:1162]: I see we're returning numbers as numbers here (rather than numeric strings). We currently return everything as strings in Search, so is this a change we're also making?

File: docs/en/sql-reference/functions/json-functions.md (hunk 14 above)
Comment (ollidraese) [reply docs/en/sql-reference/functions/json-functions.md:1112]: That could only be a string->string map, right? I think that would make sense if only a small subset of fields need to be extracted from the output and the same could be achieved by leaving the fieldname for unwanted fields empty (they don't make it to the output).
I thunk, I would like to limit this to the current functionality and look into a toMap later...

File: docs/en/sql-reference/functions/json-functions.md (hunk 15 above)
Comment (ollidraese) [reply docs/en/sql-reference/functions/json-functions.md:1162]: You can control that with the detectTypes option (defaulting to true). I would think that it would normally make sense to return the value in a native type.

File: docs/en/sql-reference/functions/json-functions.md (hunk 14 above)
Comment (cdahl-dev) [reply docs/en/sql-reference/functions/json-functions.md:1112]: I've been doing Map(String, Dynamic)

File: src/Functions/csvToJsonString.cpp [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+  * The function returns a JSON string where the field names are mapped to their corresponding values.
+  * Unless disabled, the function attempts to detect the types of values and convert them appropriately:
+  * - Numbers are converted to JSON numbers
+  * - "true" and "false" are converted to JSON booleans
+  * - All other values are treated as strings
+  * 
+  * Example:
+  * csvToJSONString('name,age', 'John,42') => '{"name":"John","age":42}'
+  * csvToJSONString('name|age', 'John|42') SETTINGS format_csv_delimiter = '|' => '{"name":"John","age":42}'
+  * csvToJSONString('name$age', 'John$42', 'delimiter="$") => '{"name":"John","age":42}'
```
Comment (itsalvinho) [inline src/Functions/csvToJsonString.cpp:45]: nit: missing single quote in `'delimiter="$"` ?

File: src/Functions/csvToJsonString.cpp [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+            auto key = Poco::toLower(keys->getDataAt(i).toString());
+            auto value = values->getDataAt(i).toString();
+
+            // remove the foramt prefix if present
+            if (key.starts_with("format_csv_"))
+            {
+                key = key.substr(11);
+            }
+
+            if (key == "delimiter" || key == "custom_delimiter")
```
Comment (itsalvinho) [inline src/Functions/csvToJsonString.cpp:178]: Can we add an option to specify the escape character? To fully support `extract type=delim`

File: src/Functions/csvToJsonString.cpp [hunk 18]
```diff
@@ -0,0 +1,408 @@
+#include <DataTypes/DataTypeNullable.h>
```
Comment (bbourbie) [inline src/Functions/csvToJsonString.cpp:1]: Super Super NIT
Usually, what I do is I put the "" includes before the <> includes.

Why??? 
For compilation performance reasons, the IO is one of the slowness part of the compilation, so I try to reduce as much as possible the includes,
It's very frequent when we include a "" , we also include a <>  (not the other way around) 

so, a trick that I do is I unclude the "" and then the <> to have the least included files.

File: src/Functions/csvToJsonString.cpp [hunk 19]
```diff
@@ -0,0 +1,408 @@
...
+#include "Columns/ColumnNullable.h"
+#include "Columns/ColumnString.h"
+#include "FunctionHelpers.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
```
Comment (bbourbie) [inline src/Functions/csvToJsonString.cpp:23]: I see it's like that CH does in other files but not so great in my opinion.. I like including a file... Anyway

File: src/Functions/csvToJsonString.cpp [hunk 20]
```diff
@@ -0,0 +1,408 @@
...
+
+    /** Splits a CSV string into a vector of strings according to the provided settings.
+      * @param s The input CSV string to split
+      * @param vectorBuffer A pre-allocated vector to store the split fields
+      * @param settings CSV format settings that control parsing behavior (delimiter, quotes, etc)
+      */
+    static void splitCSV(const StringRef & s, Strings & vectorBuffer, const FormatSettings::CSV & settings)
+    {
+        ReadBufferFromMemory buf(s.data, s.size);
+        vectorBuffer.clear();
```
Comment (bbourbie) [inline src/Functions/csvToJsonString.cpp:60]: FYI, I check ReadBufferFromMemory doesn't have a move constructor.

File: src/Functions/csvToJsonString.cpp [hunk 21]
```diff
@@ -0,0 +1,408 @@
...
+      * @param detectTypes A boolean flag to determine if types should be detected and converted
+      */
+    static void mergeAsJSON(
+        const Strings & fieldNames,
+        const Strings & fieldValues,
+        const FormatSettings::CSV & settings,
+        std::stringstream & dest,
+        bool detectTypes)
+    {
+        const auto n = std::min(fieldNames.size(), fieldValues.size());
```
Comment (bbourbie) [inline src/Functions/csvToJsonString.cpp:88]: Can the 2 sizes be different?? I guess yes when the data is garbage in, when we provide more data than col?
Don't we want to trace it we have a problem here?
because the extra data will be lost silently

File: src/Functions/csvToJsonString.cpp [hunk 22]
```diff
@@ -0,0 +1,408 @@
...
+            if (value == settings.null_representation)
+            {
+                json.set(key, Poco::Dynamic::Var()); // Sets null value
+            }
+            else
+            {
+                if (detectTypes)
+                {
+                    double num;
+                    if (Poco::NumberParser::tryParseFloat(value, num))
```
[the remark this answers is not in the document]
Comment (bbourbie) [reply src/Functions/csvToJsonString.cpp:110]: I have tested, yes, it's the case.
if I do
```
CREATE TABLE csv_to_json_test (entry Int32, csv String) ENGINE = Memory;

INSERT INTO csv_to_json_test
VALUES
    (1, 'Key1,"42",John'),
    (2, 'Key2,true,Bob'),
    (3, 'Key3,false,Alice'),
    (4, 'Key4,something,Charlie'),
    (5, 'Key5|17|Rudolf'),
    (6, '"The,key,with""quotes","\\N",",;''"');

-- with default and with SETTINGS override
SELECT 1 as T, csvToJSONString ('id,property,name', csv) AS json FROM csv_to_json_test WHERE entry < 5 ORDER BY entry;
```

we can see that the "42" string is converted to 42 number if even we try to parse in float

Also, why don't we try first integer? we try if there's any period in the string?

**Review by bbourbie — APPROVED** [review]

File: src/Functions/csvToJsonString.cpp (hunk 19 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:23]: I think, we should follow the CH pattern (in their code) as much as possible...

File: src/Functions/csvToJsonString.cpp [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+    }
+
+    /** Merges a vector of field names and values into a JSON String.
+      * @param fieldNames A vector of field names
+      * @param fieldValues A vector of field values
+      * @param settings CSV format settings to get the null representation
+      * @param dest A reference to the destination stringstream to store the JSON result
+      * @param detectTypes A boolean flag to determine if types should be detected and converted
+      */
+    static void mergeAsJSON(
```
[the remark this answers is not in the document]
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:81]: Creating string streams has significant initialization costs, so reusing them in loops (here one per row) is super important. Here is the summary from ChatGPT:

The overhead of creating a `std::stringstream` inside a tight loop compared to reusing it and resetting its contents with `str("")` can be non-trivial, particularly in performance-critical or high-iteration scenarios. Below is a detailed breakdown.

---

## 1. **Recreating vs. Reusing: Core Differences**

### **Recreating `std::stringstream` in Each Iteration**
```cpp
for (int i = 0; i < N; ++i) {
    std::stringstream ss;
    ss << some_data;
    // use ss.str()
}
```
- **Construction and destruction cost**: Each iteration allocates memory for a new internal buffer (typically via `std::stringbuf`).
- **Heap allocations**: Even though small buffer optimizations (SBO) may avoid heap allocations for short strings, repeated construction/destruction stresses allocator behavior and may lead to more frequent allocations and deallocations.
- **Cache locality**: Poorer compared to reusing one instance—new objects might not benefit from cache warmth.

### **Reusing `std::stringstream`**
```cpp
std::stringstream ss;
for (int i = 0; i < N; ++i) {
    ss.str("");       // Clears the contents
    ss.clear();       // Resets state flags (e.g., eofbit)
    ss << some_data;
    // use ss.str()
}
```
- **No allocation per iteration** (after initial buffer grows to max seen size).
- **Better memory reuse and locality**.
- **Clearing cost**: `str("")` resets the buffer but does not deallocate it, and `clear()` resets stream state.

---

## 2. **Performance Considerations**

### **Microbenchmark Data**
While exact performance depends on the standard library implementation and platform, benchmarks typically show:

| Operation                    | Relative Cost   |
|-----------------------------|-----------------|
| Constructing stringstream   | High            |
| `str("")` + `clear()`       | Low             |
| Heap allocations per loop   | Avoided if reused |
| Buffer reallocation         | Avoided if reused |

In practice:
- Recreating may be **2x–10x slower** depending on I/O volume, buffer sizes, and loop count.
- In tight loops (`10^6+` iterations), this adds measurable CPU time and allocation pressure.

---

## 3. **When to Recreate**
Recreation may be justified when:
- The stream is used briefly and infrequently.
- Reuse risks thread-safety issues (e.g., parallelism without per-thread storage).
- State cleanup for complex stream manipulations becomes too error-prone.

---

## 4. **Recommendation**
Prefer **reuse with `str("")` and `clear()`** for:
- Performance-critical paths
- Large iteration counts
- Streams with consistent usage patterns

Ensure proper state reset (`clear()`) before reuse, especially if `failbit`, `eofbit`, etc., were set in prior operations.

---

Would you like a benchmark comparison with example code?

File: src/Functions/csvToJsonString.cpp [hunk 24]
```diff
@@ -0,0 +1,408 @@
...
+                    }
+                }
+                else
+                {
+                    json.set(key, value);
+                }
+            }
+        }
+
+        dest.str(""); // reset std::stringstream
```
[the remark this answers is not in the document]
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:134]: Similar response here... and output of ChatGPT again :-)

Yes, there is a **minor performance difference** between clearing a `std::stringstream` with `.str(std::string())` versus `.str("")`, though in most cases, this difference is negligible unless you're in a **very tight loop** or **allocating large strings**.

---

## 1. **What Both Methods Do**

### `.str("")`
- Creates a temporary `const char*` string literal.
- Internally constructs a `std::string` from that literal.
- Passes a small string (typically using **Short String Optimization**, or **SSO**).
- **Avoids heap allocation** entirely (on most implementations).

### `.str(std::string())`
- Creates a temporary, **default-constructed** `std::string` (i.e., an empty `std::string`).
- Also likely SSO-optimized.
- May result in slightly different codegen depending on the compiler, but typically also **avoids heap allocation**.

---

## 2. **Performance Impact**

### **Instruction Overhead**
- `str("")` is marginally faster due to avoiding a call to the `std::string` constructor.
  - The literal is directly passed and a small SSO-enabled string is created in-place.
- `str(std::string())` may result in a **default construction followed by a copy/move**, although compilers often optimize this away.

### **Memory Considerations**
Both methods result in the internal buffer of the stringstream being reset to empty.
- Neither method will deallocate the buffer already held by the `stringstream` unless it's swapped or destructed.
- **Neither method reclaims memory** if the buffer previously held a large string. If memory reclamation is required, you'd need to **swap with a new stringstream**.

---

## 3. **Benchmark Insight (Micro)**

In practice:
- On modern compilers with optimization, both compile down to **very similar or identical code**.
- `str("")` may outperform by **a few nanoseconds per call** due to less temporary object construction.

---

## 4. **Best Practice**

Use `.str("")`:
- It's idiomatic and slightly more efficient.
- It is semantically clearer for "clear/reset to empty".
- It minimizes unnecessary object creation.

Use `.str(std::string())` only if:
- You’re reusing a preconstructed empty `std::string`, e.g., `ss.str(emptyStr)` in a loop.

---

## Summary

| Method               | Construction Overhead | Heap Allocation | Clarity | Performance (Best) |
|----------------------|------------------------|------------------|---------|---------------------|
| `.str("")`           | Minimal                | No (SSO)         | High    | ✅                  |
| `.str(std::string())`| Slightly more          | No (SSO)         | Moderate| Slightly slower     |

Unless you are profiling at nanosecond scale or relying on specific memory semantics, **prefer `.str("")` for clearing**.

Would you like compiler-specific disassembly comparison for further validation?

File: src/Functions/csvToJsonString.cpp (hunk 17 above)
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:178]: That would be a major change to the CH CSV Reader helper that I would like to avoid for now. First, we should check if anybody is actually using the option as the CSV standard tells to use "" to escape ".

File: src/Functions/csvToJsonString.cpp [hunk 25]
```diff
@@ -0,0 +1,408 @@
...
+            // csv value also provided as a string constant?
+            if (const auto * constCSV = checkAndGetColumnConst<ColumnString>(arguments[1].column.get()); constCSV)
+            {
+                splitCSV(constCSV->getValue<String>(), vectorBuffer, settings);
+                mergeAsJSON(fieldNames, vectorBuffer, settings, dest, detectTypes);
+                return DataTypeString().createColumnConst(input_rows_count, dest.str());
+            }
+
+            // csv value provided as a string (or nullable string) column
+            if (const auto * colCSV = checkAndGetColumn<ColumnString>(removeNullable(arguments[1].column).get()))
```
[the remark this answers is not in the document]
Comment (ollidraese) [reply src/Functions/csvToJsonString.cpp:356]: we do not need to test for colCSV as it is guaranteed to be a ColumnString after removing the nullable
